### PR TITLE
DOMA-4185 Added new InviteNewOrganizationEmployees access

### PR DESCRIPTION
**Pull request by VKislov**
Separated the accesses of the employee invite service to the organization from the `OrganizationEmployee` accesses.

### CODE REVIEW

File: apps/condo/domains/organization/access/OrganizationEmployee.js [hunk 1]
```diff
@@ -73,8 +74,12 @@ async function canManageOrganizationEmployees ({ authentication: { item: user },
             organization: { id: employeeToEdit.organization },
         })
         if (!employeeRole) return false
+        let updatePermission = await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, employeeToEdit.organization, 'canUpdateEmployees')
```
Comment (akarjakin) [inline apps/condo/domains/organization/access/OrganizationEmployee.js:77]:
```
const accessType = originalInput.deletedAt ? 'canDeleteEmployees' : 'canUpdateEmployees'
const updatePermission = await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, employeeToEdit.organization, accessType)
```

File: apps/condo/migrations/20221013103705-0177_organizationemployeerole_cancreateemployees_and_more.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+--
+-- Add field canDeleteEmployees to organizationemployeerolehistoryrecord
+--
+ALTER TABLE "OrganizationEmployeeRoleHistoryRecord" ADD COLUMN "canDeleteEmployees" boolean NULL;
+--
+-- Add field canUpdateEmployees to organizationemployeerolehistoryrecord
+--
+ALTER TABLE "OrganizationEmployeeRoleHistoryRecord" ADD COLUMN "canUpdateEmployees" boolean NULL;
+
+-- set canCreateEmployees to true and canManageEmployees to false for Foreman employee role
```
Comment (akarjakin) [inline apps/condo/migrations/20221013103705-0177_organizationemployeerole_cancreateemployees_and_more.js:35]: 1. no need for 3 updates, 1 should be enough
2. use CAPS for SQL commands like SELECT, SET, WHERE, ...
3. no DOWN migration updates reverting these changes
4. looks like this updates should be moved to scripts.

File: apps/condo/domains/organization/schema/OrganizationEmployeeRole.js [hunk 3]
```diff
@@ -52,6 +52,9 @@ const OrganizationEmployeeRole = new GQLListSchema('OrganizationEmployeeRole', {
         canManageIntegrations: { type: Checkbox, defaultValue: false },
         canManageProperties: { type: Checkbox, defaultValue: false },
         canManageTickets: { type: Checkbox, defaultValue: false },
+        canCreateEmployees: { type: Checkbox, defaultValue: false },
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/OrganizationEmployeeRole.js:55]: Use canManageEmployees for crud actions

**Review by pahaz — CHANGES_REQUESTED** [review]

File: apps/condo/domains/organization/access/OrganizationEmployee.js [hunk 4]
```diff
@@ -48,8 +49,8 @@ async function canManageOrganizationEmployees ({ authentication: { item: user },
             id: employeeForUser.role,
             organization: { id: employeeForUser.organization },
         })
-
-        return !!(employeeRole && employeeRole.canManageEmployees)
+        const createPermission = await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, organizationId, 'canCreateEmployees')
+        return !!(employeeRole && (employeeRole.canManageEmployees || createPermission))
```
Comment (SavelevMatthew) [inline apps/condo/domains/organization/access/OrganizationEmployee.js:53]: Why old `canManage` and `canCreate` at the same time?

File: apps/condo/pages/employee/[id]/index.tsx [hunk 5]
```diff
@@ -277,7 +277,7 @@ export const EmployeeInfoPage = () => {
     const updateEmployeeAction = OrganizationEmployee.useUpdate({}, () => refetch())
     const softDeleteAction = OrganizationEmployee.useSoftDelete(() => Router.push('/employee/'))
 
-    const isEmployeeEditable = canManageEmployee(link, employee)
+    const isEmployeeEditable = get(link, ['role', 'canManageEmployees'], null) || get(link, ['role', 'canUpdateEmployees'], null)
```
Comment (SavelevMatthew) [inline apps/condo/pages/employee/[id]/index.tsx:280]: It's a bit strange to keep old and new format at the same time...
I'm 100% sure someone else will forget about it later...

So I guess you need to migrate old format to new or smth...

File: apps/condo/domains/organization/access/OrganizationEmployee.js [hunk 6]
```diff
@@ -48,8 +49,8 @@ async function canManageOrganizationEmployees ({ authentication: { item: user },
             id: employeeForUser.role,
             organization: { id: employeeForUser.organization },
         })
-
-        return !!(employeeRole && employeeRole.canManageEmployees)
+        const createPermission = await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, organizationId, 'canCreateEmployees')
```
Comment (AntonAL) [inline apps/condo/domains/organization/access/OrganizationEmployee.js:52]: A name `createPermission` looks like an action

```suggestion
        const permissionOnCreate = await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, organizationId, 'canCreateEmployees')
```

File: apps/condo/domains/organization/access/OrganizationEmployee.js [hunk 7]
```diff
@@ -73,8 +74,10 @@ async function canManageOrganizationEmployees ({ authentication: { item: user },
             organization: { id: employeeToEdit.organization },
         })
         if (!employeeRole) return false
+        const accessType = originalInput.deletedAt ? 'canDeleteEmployees' : 'canUpdateEmployees'
+        const updatePermission = await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, employeeToEdit.organization, accessType)
```
Comment (AntonAL) [inline apps/condo/domains/organization/access/OrganizationEmployee.js:78]:
```suggestion
        const permissionOnUpdate = await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, employeeToEdit.organization, accessType)
```

**Comment by VKislov** [comment]
Rebased and added UPDATE canManageEmployee for Manager role

File: apps/condo/domains/organization/access/InviteNewOrganizationEmployeeService.js [hunk 8]
```diff
@@ -1,16 +1,33 @@
...
+    if (!employeeForUser) {
+        return false
+    }
+
+    const employeeRole = await getByCondition('OrganizationEmployeeRole', {
+        id: employeeForUser.role,
+        organization: { id: employeeForUser.organization },
+    })
+    const permissionOnCreate = await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, organizationId, 'canInviteNewOrganizationEmployees')
+    return !!(employeeRole && permissionOnCreate)
```
Comment (pahaz) [inline apps/condo/domains/organization/access/InviteNewOrganizationEmployeeService.js:30]:
```suggestion
    return await checkPermissionInUserOrganizationOrRelatedOrganization(user.id, organizationId, 'canInviteNewOrganizationEmployees')
```

**Review by pahaz — APPROVED** [review]

**Pull request merged**